Adds the block hierarchy navigation menu to the header

youknowriad opened the pull request:
closes #9628 and closes #9212

This is still WIP, there are some things we need to polish before landing this:

- [x] The design is ready.
- [x] Add some unit tests for the new selectors and an e2e test for the navigation
- [x] Fix https://github.com/WordPress/gutenberg/issues/9212 which is causing the inner blocks to be selected when we select the parent.

Doesn't mean we can't start testing and reviewing the PR though.

youknowriad committed:
Adds the block hierarchy navivation menu to the header

mtias commented:
"Styles" should probably be displayed at the top.

mtias commented:
This is working quite well for me. Agreed on needing to fix #9212. We should also update the single column block icon.

mtias commented:
Currently:

![image](https://user-images.githubusercontent.com/548849/46935833-a2b2de00-d05c-11e8-9ecb-a8527076f123.png)

youknowriad committed:
Avoid auto-selecting inner blocks when selecting the parent block

youknowriad committed:
Show block hierarchy for container blocks

youknowriad committed:
highlight the selected block

youknowriad committed:
Add a unit test for the newly added selector

youknowriad committed:
Adding an e2e test

youknowriad committed:
Replace block names by block titles

youknowriad committed:
Fix unit tests

youknowriad committed:
Fix e2e tests

youknowriad commented:
I fixed the selection bug, I added some tests. The technical pieces should be ready. We might want to do some design tweaks now (cc @jasmussen ) before landing.

youknowriad commented:
Oh I think @jasmussen is AFK this week (sorry for the pings), I'll see what we can do here :).

youknowriad committed:
Update the selector docs

youknowriad committed:
Tweak the design of the block navigation menu

youknowriad commented:
I tried tweaking the design as much as I can. Let me know how I should polish it further. and yes, this is ready for a final review.

Soean commented:
Currently:
![bildschirmfoto 2018-10-15 um 13 13 54](https://user-images.githubusercontent.com/695201/46947776-4364c600-d07c-11e8-9df9-f1beb8702622.png)

mtias commented:
> I tried tweaking the design as much as I can. Let me know how I should polish it further.

Let's try to compress it a little so that the second level doesn't come so far into the right.

youknowriad committed:
Compress the block hierarchy tree

youknowriad commented:
<img width="260" alt="screen shot 2018-10-15 at 12 37 55" src="https://user-images.githubusercontent.com/272444/46948855-59bc5300-d077-11e8-82d2-a1ceef4b5654.png">

Compressed, here's a multi-level tree

mtias committed:
Update inner column block icon.

mtias commented:
Looking good, updated the icon.

youknowriad committed:
Avoid showing the parent block's borders if a child block is selected

youknowriad committed:
Simplify the menu's label

mtias commented:
Final thing: can we use a background on the "selected" block element?

![image](https://user-images.githubusercontent.com/548849/46951498-6e551700-d088-11e8-9157-36a4378888bf.png)

tofumatt commented:
Yes please to the selected block.

youknowriad committed:
Tweak the selected block's style

youknowriad commented:
Updated

<img width="260" alt="screen shot 2018-10-15 at 13 49 14" src="https://user-images.githubusercontent.com/272444/46951878-2252a400-d081-11e8-97ca-562e317475ba.png">

mtias commented:
Might need to use a darker gray for the text on the selected item.

youknowriad committed:
Darker text color for the selected block

youknowriad commented:
What about this?

<img width="261" alt="screen shot 2018-10-15 at 13 55 20" src="https://user-images.githubusercontent.com/272444/46952181-0bf91800-d082-11e8-85e9-e2e77d2dc1c2.png">

aduth commented on packages/editor/src/store/selectors.js:
```diff
@@ -782,6 +782,26 @@ export function getBlockRootClientId( state, clientId ) {
+ *
+ * @return {string} Root client ID
+ */
+export function getBlockHierarchyRootClientId( state, clientId ) {
```
This in combination with `getBlockRootClientId` isn't very optimized, particularly when a consuming UI is always present on the screen. Mentioning because I've seen it come up already as one of the higher time-wasters in performance audit results. I might think we should try to do one of:

- Avoid calling the selector if possible (e.g. only when dropdown open)
- Memoize one or both selectors
- Reshape state to optimize upward-path traversal
   - This one is actually fairly straight-forward with adding a `rootClientId` property to the block state. I've only been reluctant because it's technically redundant, but I think performance would win out here.

aduth commented on packages/editor/src/components/block-list/block.js:
```diff
@@ -131,7 +133,12 @@ export class BlockListBlock extends Component {
 
 		// Find all tabbables within node.
-		const textInputs = focus.tabbable.find( this.node ).filter( isTextField );
+		const innerBlocks = this.node.querySelector( '.editor-inner-blocks' );
```
This is a future bug waiting to happen.

youknowriad replied:
How would you solve it?

aduth replied:
I worked my way back to questioning whether we need it at all in following comments. We can revisit if still then necessary.

aduth replied:
We have an `isInSameBlock` function in `editor/utils/dom.js` which could work well for this.

youknowriad replied:
@aduth It won't work for the column's block appender. As it's in the same block but we don't want to focus it.

aduth replied:
Seems like a problem there is that we consider the default appender as a valid text field to target? For what reason is it `<input type="text">`? Should we exclude `readOnly` from auto-focus?

aduth replied:
I'd be content if we at least abstract this to some common utility in `editor/utils/dom.js`. The file was created after this same sort of bugginess crept up repeatedly, to have a single point of updates for DOM inspection functions.

e.g. `isInSameRootClientId`

youknowriad replied:
The appender is an input to allow the writing flow navigation (we discussed this at some point :) I think).

I extracted to a dom utility.

aduth commented on packages/editor/src/components/block-list/block.js:
```diff
@@ -91,7 +91,7 @@ export class BlockListBlock extends Component {
 
 		if ( this.props.isSelected && ! prevProps.isSelected ) {
-			this.focusTabbable();
+			this.focusTabbable( true );
```
What is it we're trying to do here? And why is it inconsistent with the behavior in `componentDidMount`?

aduth replied:
Okay, so if it's to avoid "select" actually causing one of its inner blocks to become selected, can't this also happen in a newly-mounted block as well?

aduth replied:
Maybe we should reconsider whether to call `focusTabbable` on a selection update as well. I don't recall what we need it for.

youknowriad replied:
> Maybe we should reconsider whether to call focusTabbable on a selection update as well. I don't recall what we need it for.

Initial implementation here was to avoid selecting innerBlocks in all cases but this breaks the writing flow when inserting columns (it actually broke 2 e2e tests) because the first paragraph was not focused in that case. I did consider whether it's fine to not focus the first paragraph in that case as assume we hit "tab" or click no the appender of the column block first but I think from a "Writing flow" perspective, it's better if we focus the first paragraph in a column directly when creating a columns block.

aduth replied:
>it's better if we focus the first paragraph in a column directly when creating a columns block.

Personally I've found this behavior to be a weird / unexpected side effect of focus behaviors. On the Layout<->Writing spectrum, Columns falls more on the former, so I don't necessarily see the paragraph creation / focus being a necessity.

youknowriad replied:
> As "Columns" evolve, it'd be natural to focus on the container when it is added rather than on the paragraph within.

Yes one idea I had was to rewrite the columns block similarly to the table block where you select the number of columns before typing but I though this was out of the scope of this PR anyway.

aduth replied:
I'm getting a sense we're becoming very ad hoc with how and when we expect focus transitions to occur upon block selection, hacking in various conditions to preserve intended flows while losing sight of what those intended flows are. It's signalling to me that naively checking for / comparing on `isSelected` from `BlockListBlock` is not serving us very well, and that perhaps it ought to be moved / refactored to more appropriately capture the user interactions where we expect it to occur: as best I can tell, this is mostly in splitting / merging / replacing actions occurring while typing within a `RichText`.

youknowriad commented:
> Also: would love to see how this works with a really big block tree. What's the scrolling situation?

It works pretty well since the popover is designed to adapt to the screen and scroll if necessary.

> Is that enough contrast for a11y?

It's waiting for a review

chrisvanpatten commented:
@youknowriad Thanks, missed that tag :)

youknowriad committed:
Memoize the block parent's selectors

karmatosed reviewed (approved):
Approving design. Thanks for this.

tofumatt reviewed (dismissed):
So this is really cool and keyboard navigation *within* the modal works great. You can use <kbd>Space</kbd> or <kbd>Enter</kbd> to activate a block and navigate to it. Super cool and makes for *excellent* block nav. Adding nicer focus styles could be cool, if they are the right contrast.

**However**, there's not actually a way, that I can see, to *access* the toolbar item with the keyboard. Because right now it's contextual (it only shows for blocks with `innerBlocks`, and only shows when they're activated), I can't tab to it. Tabbing to it also takes a long time. A keyboard shortcut would be good, but I also think this should always be visible and should show just root blocks in order.

If that's the case we could name it "Block Navigation" and we could address #5694.

So I think:

1. this should have a keyboard shortcut
2. it should always appear, not contextually appear
3. it should show all blocks

I hope that's not a pain to implement, but if we did it it'd be soooooooooo good for keyboard users.

tofumatt commented:
Oh, I forgot to mention: it passes colour contrast.

<img width="447" alt="screenshot 2018-10-15 at 20 10 37" src="https://user-images.githubusercontent.com/90871/46972424-657b3a00-d0b6-11e8-83ac-e721749a1a43.png">

👍

youknowriad committed:
Adding root level block navigation

youknowriad committed:
Adding a keyboard shortcut to open the navigation menu

youknowriad committed:
Avoid wrapping function descriptions

gziolo commented on packages/editor/src/components/block-navigation/dropdown.js:
```diff
@@ -0,0 +1,36 @@
+
+const menuIcon = (
+	<SVG xmlns="http://www.w3.org/2000/svg" viewBox="0 0 24 24" width="20">
+		<path d="M5 5H3v2h2V5zm3 8h11v-2H8v2zm9-8H6v2h11V5zM7 11H5v2h2v-2zm0 8h2v-2H7v2zm3-2v2h11v-2H10z" />
```
Can you use `Path` component instead of HTML tag? We probably should refactor all code occurrences and add linter rule for that.

aduth replied:
>add linter rule for that.

I did the hard work:

https://astexplorer.net/#/gist/f7f5d60c5bbc381708bff5a035cfe1ad/beb956688752e5b3cf616e77c7b153130b3b979b

youknowriad replied:
Thanks Andrew, 

<img width="297" alt="screen shot 2018-10-16 at 15 57 52" src="https://user-images.githubusercontent.com/272444/47025911-46cf7e80-d15c-11e8-8bf6-a47f1faf763f.png">

I'll open a separate PR, as there's still some work to do :P

youknowriad replied:
Should we also use the `G` cc @gziolo ?

youknowriad replied:
What about rect...?

youknowriad committed:
Use Path for a mobile-friendly SVG

youknowriad committed:
Extract inner blocks check to a dom utility

youknowriad committed:
Fix unit tests

aduth commented:
>However, there's not actually a way, that I can see, to access the toolbar item with the keyboard. 

It is possible with the <kbd>Ctrl + `</kbd> region navigation:

https://cloudup.com/cSZe9jH9eWJ

youknowriad committed:
Fix e2e tests

mtias commented:
We cannot use it in the browser, but `<kbd>Command + T</kbd>` would be nice as navigation.

tofumatt commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -44,6 +44,10 @@ const globalShortcuts = {
 			ariaLabel: shortcutAriaLabel.primaryShift( ',' ),
 		},
+		{
+			keyCombination: [ 'Esc', 'Esc' ],
```
Turns out this keyboard command might be a bit too fast for some users. I noticed to get it to happen I had to press <kbd>Esc</kbd> twice in a row *pretty quickly*. I'm not if that's an OS-level setting (I'm guessing it's something our keyboard shortcut handler defines), but at its current speed it's *way* too tight a tolerance for users with mobility issues.

We should probably find another shortcut for this. I know there are so few, but we'll have to 🤷‍♂️

I'm open to suggestions and I'll try to come up with one.

tofumatt replied:
Possible keyboard options I've tried thus far:

* <del><kbd>⌥⌘B</kbd></del> (Used by Chrome for Bookmarks Bar)
* <del><kbd>⇧⌘B</kbd></del> (Used by Chrome for Bookmarks Page)

tofumatt reviewed (dismissed):
I updated VMWare earlier and it meant I had to do a dance to get my Windows VM with NVDA up again. Sorry 'bout that.

Overall this works really well.

The main issue here is around the keyboard shortcut. I tried using a screenreader and things worked well, including with the keyboard shortcut ("Block Navigation List" is spoken when opened with the keyboard shortcut, which is good.)

I think we need a modifier key shortcut. What do we have available?

tofumatt reviewed (commented):
Arg, sorry, I was wrong about the keyboard shortcut giving context to the list, but I have a fix for it incoming.

tofumatt committed:
Add aria-label to block navigation list to improve context with keyboard shortcut

tofumatt commented on packages/editor/src/components/block-navigation/index.js:
```diff
@@ -24,7 +25,10 @@ function BlockNavigationList( {
 	return (
-		<ul className="editor-block-navigation__list">
+		<ul
+			aria-label={ __( 'Block Navigation' ) }
```
Oh, @youknowriad pinged me and pointed out this shouldn't be here; we should move this into the block as content (an off-screen header I guess).

I used it with NVDA this way and it only read the label when I used the keyboard shortcut, not when I activated the dropdown using the toggle button. The toggle button provides context as to what this is so it seemed like noise to add it to the content.

But if it's not too annoying to read it out again after the button announcing the content we can move it in, sure 😄

youknowriad committed:
Use the label prop to label the MenuGroup

youknowriad committed:
Use access + b shortcut to avoid ambiguity

afercia commented:
> We cannot use it in the browser, but Command + T would be nice as navigation.

> I think we need a modifier key shortcut. What do we have available?

As discussed at length in https://github.com/WordPress/gutenberg/pull/3084 and especially in https://github.com/WordPress/gutenberg/issues/3218 single-modifier shortcuts are basically impossible to use because they will very likely conflict with other shortcuts. Also, not all keyboard layouts provide the same keys. A more reliable approach would be using two modifiers. However, the only final solution to this issue is a mechanism to allow users to re-map and customize shortcuts as proposed in https://github.com/WordPress/gutenberg/issues/3218 and as most of the desktop editors do.

afercia commented:
@tofumatt looking at this PR for the first time, there's a technical mistake that needs to be fixed.
Based on the ARIA specification, elements with role=menu must have direct children with role=menuitem.

This menu instead contains a list, list items contain a divs with role=presentation, which contain buttons with a role=menuitem. 

Ideally, the UL, LI, and DIV should be completely removed. Why are they even used? If it's for styling purposes, then it's not a valid reason to break the accessibility tree. CSS should be used instead of redundant markup.

Less preferably, and it would potentially expose to AT bugs, all the redundant elements need to be removed from the accessibility tree with role=presentation in order to "flatten" the tree and make AT think the buttons are children of the menu.

Currently, the tree is broken and anyone can just check it in the Chrome of Firefox dev tools:

<img width="405" alt="screen shot 2018-10-17 at 00 18 31" src="https://user-images.githubusercontent.com/1682452/47051118-3262a480-d1a3-11e8-83e3-b21e4768bed6.png">

youknowriad committed:
Add role="presentation" to the items that should be invisible a11y wise

youknowriad committed:
Fix unit tests

tofumatt reviewed (commented):
I have some changes to make and then I think this is good to go. I'll approve after I push my changes; if you want to check them out and make sure they're alright please do!

talldan commented on packages/edit-post/src/components/keyboard-shortcut-help-modal/config.js:
```diff
@@ -44,6 +44,10 @@ const globalShortcuts = {
 			ariaLabel: shortcutAriaLabel.primaryShift( ',' ),
 		},
+		{
+			keyCombination: access( 'b' ),
```
I'll just mention what I posted in slack here. I started compiling a list of browser shortcuts in a spreadsheet a while ago:
https://docs.google.com/spreadsheets/d/1nK1frKawxV7aboWOJbbslbIqBGoLY7gqKvfwqENj2yE/edit#gid=0

I note that we had access( 'l' ) earmarked for Align Left. How about access( 'o' ), as in O for outline?

tofumatt committed:
change shortcuts, improve tests, and show empty navigator when no blocks exist

tofumatt commented:
The only thing to note here is NVDA reads out "Block Navigation Block Navigation menu" when this is opened. I wonder if the `label` prop on `MenuGroup` is repeating the label somehow?

That is a bug but we can tweak it after we merge this (which we should do for UI freeze). Once this is merged either file a follow-up issue and ping me or I'll do it 😄

tofumatt committed:
Tweak tests; use o instead of l

tofumatt commented:
Changed the shortcut to "o" (see: https://wordpress.slack.com/archives/C02QB2JS7/p1539717395000100) and tweaked the tests. This gets the 👍 from me.

tofumatt reviewed (approved):
Changed the shortcut to "o" (see: https://wordpress.slack.com/archives/C02QB2JS7/p1539717395000100) and tweaked the tests. This gets the 👍 from me.

afercia commented:
Shortcuts: for future reference, worth nothing on the new WAI-ARIA Authoring Practices 1.1 (still not an official WCAG recommendation) there's a pretty good section dedicated to keyboard shortcuts. It's more about what _not_ to do: https://www.w3.org/TR/wai-aria-practices-1.1/#kbd_shortcuts_assigning

talldan commented:
I've also made a PR updating the docs with that new keyboard shortcut, would be great if someone could approve:
https://github.com/WordPress/gutenberg/pull/10677

talldan commented:
One other thing, I noticed this when testing the new menu #10679. I don't think it's a massive issue, but it did cause some weirdness.

paaljoachim commented:
I am testing 4.1 RC2
I am not totally read up on how this works but I noticed the above wireframe with the icon beside Content Structure. Opening..
<img width="298" alt="screen shot 2018-10-22 at 17 09 31" src="https://user-images.githubusercontent.com/5323259/47300631-41bb7500-d61d-11e8-8064-7811f7e144b6.png">

I see it is a way to move around in the document. I also expected to be able to move the items directly in the list. Instead of having to go to the document and move items there.

tofumatt commented:
@simbalion and @paaljoachim: Interesting ideas about expectations from these features. I actually think they'd go together; an outline of the selected block would be a great addition. And then a mover would make sense because you could see what you're moving.

Not sure about the mover but the outline I'd definitely 👍 as an idea. Could you both file new issues for the features you'd like? This PR is merged so it's not the best place to discuss additional features, but new issues would be great. Thanks!

simbalion commented:
Moving blocks would work presumably the same as it normally does, the only difference would be the visibility of each block's borders, and the padding between them. Blocks could be quickly and visibly arranged, and then the 'explode' status could be toggled off to return to WYSIWYG editing.

paaljoachim commented:
It seems this issue: https://github.com/WordPress/gutenberg/issues/7115 with Insert Summary feature is somewhat similar to the Block Navigation.
